mono/metadata/marshal.c

```diff
@@ -8169,13 +8196,39 @@ mono_marshal_emit_managed_wrapper (MonoMethodBuilder *mb, MonoMethodSignature *i
 		}
 	}
 
-	if (mono_threads_is_coop_enabled ()) {
-		leave_pos = mono_mb_emit_branch (mb, CEE_LEAVE);
+	leave_try_pos = mono_mb_emit_branch (mb, CEE_LEAVE);
 
-		/* } finally { */
-		clause->try_len = mono_mb_get_label (mb) - clause->try_offset;
-		clause->handler_offset = mono_mb_get_label (mb);
-	}
+	/* } catch (Exception e) { */
+	clause_catch->try_len = mono_mb_get_label (mb) - clause_catch->try_offset;
+	clause_catch->handler_offset = mono_mb_get_label (mb);
+
+	mono_mb_emit_stloc (mb, e_local);
+
+	/* if (!mono_marshal_has_ftnptr_eh_callback ()) { */
+	mono_mb_emit_icall (mb, mono_marshal_has_ftnptr_eh_callback);
+	rethrow_pos = mono_mb_emit_branch (mb, CEE_BRTRUE);
+
+	/* throw e; */
+	mono_mb_emit_ldloc (mb, e_local);
+	mono_mb_emit_byte (mb, CEE_THROW);
+
+	/* } [endif] */
+	mono_mb_patch_branch (mb, rethrow_pos);
+
+	/* ex = mono_gc_handle_new (e, true); */
+	mono_mb_emit_ldloc (mb, e_local);
+	mono_mb_emit_icon (mb, 1);
```

Comment:
No need to be pinned, the callback must switch to running to poke into managed memory anyways.